src/Microsoft.VisualStudio.Editor.Razor/RazorDirectiveCompletionSourceProvider.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    [System.Composition.Shared]
+    [Export(typeof(IAsyncCompletionSourceProvider))]
+    [Name("Razor directive completion provider.")]
+    [ContentType(RazorLanguage.ContentType)]
+    internal class RazorDirectiveCompletionSourceProvider : IAsyncCompletionSourceProvider
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+
+        [ImportingConstructor]
+        public RazorDirectiveCompletionSourceProvider(ForegroundDispatcher foregroundDispatcher)
+        {
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+        }
+
+        public IAsyncCompletionSource GetOrCreate(ITextView textView)
+        {
+            if (textView == null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            var razorBuffer = textView.BufferGraph.GetRazorBuffers().FirstOrDefault();
+
+            if (razorBuffer == null)
+            {
+                // Razor buffer hasn't been associated with the text view yet.
+                return null;
+            }
+
+            if (!razorBuffer.Properties.TryGetProperty(typeof(VisualStudioRazorParser), out VisualStudioRazorParser parser))
+            {
+                // Parser hasn't been associated with the text buffer yet.
+                return null;
+            }
```

Comment:
Is this a known possible case or just a guard?

Reply:
It guards against pre-core Razor. This wont be an issue once things move over to 15.8.